core/chains/evm/txmgr/stuck_tx_detector.go

```diff
@@ -0,0 +1,381 @@
+	httpClient := &http.Client{Transport: t}
+	return &stuckTxDetector{
+		lggr:             lggr,
+		chainID:          chainID,
+		chainType:        chainType,
+		cfg:              cfg,
+		gasEstimator:     gasEstimator,
+		txStore:          txStore,
+		chainClient:      chainClient,
+		httpClient:       httpClient,
+		purgeBlockNumMap: make(map[common.Address]int64),
+	}
+}
+
+func (d *stuckTxDetector) LoadPurgeBlockNumMap(ctx context.Context, addresses []common.Address) error {
+	if !d.cfg.AutoPurgeStuckTxs() {
+		return nil
+	}
+	d.purgeBlockNumLock.Lock()
+	defer d.purgeBlockNumLock.Unlock()
+	// Ok to reset the map here since this method could be reloaded with a new list of from addresses
+	d.purgeBlockNumMap = make(map[common.Address]int64)
+	for _, address := range addresses {
+		d.purgeBlockNumMap[address] = 0
+	}
+
+	// Find all fatal error transactions to see if any were from previous purges to properly set the map
+	txs, err := d.txStore.FindTxsByStateAndFromAddresses(ctx, addresses, txmgr.TxFatalError, d.chainID)
+	if err != nil {
+		return fmt.Errorf("failed to query fatal error transactions from the txstore: %w", err)
+	}
+
+	// Set the purgeBlockNumMap with the receipt block num of purge attempts
+	for _, tx := range txs {
+		for _, attempt := range tx.TxAttempts {
+			if attempt.IsPurgeAttempt && len(attempt.Receipts) > 0 {
+				// There should only be 1 receipt in an attempt for a transaction
+				d.purgeBlockNumMap[tx.FromAddress] = attempt.Receipts[0].GetBlockNumber().Int64()
+				break
+			}
+		}
+	}
+
+	return nil
+}
+
+// If the AutoPurgeStuckTxs feature is enabled, finds terminally stuck transactions
+// Uses a chain specific method for detection, or if one does not exist, applies a general heuristic
+func (d *stuckTxDetector) DetectStuckTransactions(ctx context.Context, enabledAddresses []common.Address, blockNum int64) ([]Tx, error) {
+	if !d.cfg.AutoPurgeStuckTxs() {
+		return nil, nil
+	}
+	txs, err := d.FindPotentialStuckTxs(ctx, enabledAddresses)
+	if err != nil {
+		return nil, fmt.Errorf("failed to find potential stuck transactions: %w", err)
+	}
+	// No transactions found
+	if len(txs) == 0 {
+		return nil, nil
+	}
+
+	var stuckTxs []Tx
+	switch d.chainType {
+	case config.ChainScroll:
+		stuckTxs, err = d.detectStuckTransactionsScroll(ctx, txs)
+	case config.ChainZkEvm:
+		stuckTxs, err = d.detectStuckTransactionsZkEVM(ctx, txs)
+	default:
+		stuckTxs, err = d.detectStuckTransactionsHeuristic(ctx, txs, blockNum)
+	}
+
+	for _, stuckTx := range stuckTxs {
+		lggr := stuckTx.GetLogger(d.lggr)
+		lggr.Debugw("marking transaction as terminally stuck", "etx", stuckTx)
+	}
+
+	return stuckTxs, err
+}
+
+// Finds the lowest nonce Unconfirmed transaction for each enabled address
+// Only the earliest transaction can be considered terminally stuck. All others may be valid and just stuck behind the nonce
+func (d *stuckTxDetector) FindPotentialStuckTxs(ctx context.Context, enabledAddresses []common.Address) ([]Tx, error) {
```

Comment:
rename this to:
findUnconfirmedTxWithLowestNonce()

Because this logic has nothing to do with stuck txs.

Also, for this struct, don't export any methods that are used internally only.

Reply:
I was exporting this one primarily for testing